client/rpcdataset.go

```diff
@@ -0,0 +1,545 @@
+				*t = math.Float64frombits(bits)
+			case *string:
+				bits := binary.BigEndian.Uint64(valueBytes)
+				*t = float64ToString(math.Float64frombits(bits))
+			default:
+				return fmt.Errorf("dest[%d] types must be *float64 or *string", i)
+			}
+		case TEXT:
+			switch t := d.(type) {
+			case *string:
+				*t = string(valueBytes)
+			default:
+				return fmt.Errorf("dest[%d] types must be *string", i)
+			}
+		default:
+			return nil
+		}
+	}
+	return nil
+}
+
+func (s *IoTDBRpcDataSet) getFloat(columnName string) float32 {
+	if s.closed {
+		return 0
+	}
+	columnIndex := s.getColumnIndex(columnName)
+	if !s.isNull(int(columnIndex), s.rowsIndex-1) {
+		s.lastReadWasNull = false
+		bits := binary.BigEndian.Uint32(s.values[columnIndex])
+		return math.Float32frombits(bits)
+	}
+	s.lastReadWasNull = true
+	return 0
+}
+
+func (s *IoTDBRpcDataSet) getDouble(columnName string) float64 {
+	if s.closed {
+		return 0
+	}
+	columnIndex := s.getColumnIndex(columnName)
+
+	if !s.isNull(int(columnIndex), s.rowsIndex-1) {
+		s.lastReadWasNull = false
+		bits := binary.BigEndian.Uint64(s.values[columnIndex])
+		return math.Float64frombits(bits)
+	}
+	s.lastReadWasNull = true
+	return 0
+}
+
+func (s *IoTDBRpcDataSet) getInt32(columnName string) int32 {
+	if s.closed {
+		return 0
+	}
+	columnIndex := s.getColumnIndex(columnName)
+	if !s.isNull(int(columnIndex), s.rowsIndex-1) {
+		s.lastReadWasNull = false
+		return bytesToInt32(s.values[columnIndex])
+	}
+
+	s.lastReadWasNull = true
+	return 0
+}
+
+func (s *IoTDBRpcDataSet) getInt64(columnName string) int64 {
+	if s.closed {
+		return 0
+	}
+	if columnName == TimestampColumnName {
+		return bytesToInt64(s.time)
+	}
+
+	columnIndex := s.getColumnIndex(columnName)
+	bys := s.values[columnIndex]
+
+	if !s.isNull(int(columnIndex), s.rowsIndex-1) {
+		s.lastReadWasNull = false
+		return bytesToInt64(bys)
+	}
+	s.lastReadWasNull = true
+	return 0
+}
+
+func (s *IoTDBRpcDataSet) hasCachedResults() bool {
+	if s.closed {
+		return false
+	}
+	return (s.queryDataSet != nil && len(s.queryDataSet.Time) > 0)
+}
+
+func (s *IoTDBRpcDataSet) next() (bool, error) {
+	if s.closed {
+		return false, errClosed
+	}
+
+	if s.hasCachedResults() {
+		s.constructOneRow()
+		return true, nil
+	}
+	if s.emptyResultSet {
+		return false, nil
+	}
+
+	r, err := s.fetchResults()
+	if err == nil && r {
+		s.constructOneRow()
+		return true, nil
+	}
+	return false, nil
+}
+
+func (s *IoTDBRpcDataSet) fetchResults() (bool, error) {
+	if s.closed {
+		return false, errClosed
+	}
+	s.rowsIndex = 0
+	req := rpc.TSFetchResultsReq{s.sessionId, s.sql, s.fetchSize, s.queryId, true}
+	resp, err := s.client.FetchResults(context.Background(), &req)
+	if err != nil {
+		return false, err
+	}
+	if !resp.HasResultSet {
+		s.emptyResultSet = true
+	} else {
+		s.queryDataSet = resp.GetQueryDataSet()
+	}
+	return resp.HasResultSet, nil
+}
+
+func (s *IoTDBRpcDataSet) IsClosed() bool {
+	return s.closed
+}
+
+func (s *IoTDBRpcDataSet) Close() error {
```

Comment:
The close function should send a close request to server to release the resource in server. You can refer to the close function in `IoTDBRpcDataSet` class in java

Reply:
PTAL